Tests/scripts/stop_running_pipeline.py

```diff
@@ -0,0 +1,111 @@
+GITLAB_CANCEL_TOKEN = os.getenv('GITLAB_CANCEL_TOKEN', '')
+CI_COMMIT_BRANCH = os.getenv('CI_COMMIT_BRANCH', '')
+CI_PIPELINE_ID = os.getenv('CI_PIPELINE_ID', '')
+GITLAB_STATUSES_TO_CANCEL = {"created", "waiting_for_resource", "preparing", "pending", "running"}
+
+
+def options_handler() -> argparse.Namespace:
+    parser = argparse.ArgumentParser(description='Stop running pipelines')
+    parser.add_argument('-u', '--url', help='The gitlab server url', default=GITLAB_SERVER_URL)
+    parser.add_argument('-gp', '--gitlab-project-id', help='The Gitlab project id', default=GITLAB_PROJECT_ID)
+    parser.add_argument('-c', '--ci-token', help='The token for Gitlab', required=False, default=GITLAB_CANCEL_TOKEN)
+    parser.add_argument('--pipeline-id', help='Current Pipeline ID', required=False, default=CI_PIPELINE_ID)
+    parser.add_argument('--current-branch', required=False, help='Current branch name', default=CI_COMMIT_BRANCH)
+    return parser.parse_args()
+
+
+def branch_name_for_test_upload_flow(branch_name: str, pipeline_id: str) -> str:
+    return f"{branch_name}{BUCKET_UPLOAD_BRANCH_SUFFIX}-{pipeline_id}"
+
+
+def get_all_pipelines_for_all_statuses(project: Project, branch_name: str, triggering_source: str) -> list[ProjectPipeline]:
+    # Get all pipelines for all statuses, some pipelines might be duplicated as they might change
+    # transition (e.g. from pending to running)
+    pipelines: dict[str, ProjectPipeline] = {}
+    for status in GITLAB_STATUSES_TO_CANCEL:
+        pipelines_for_status: list[ProjectPipeline] = project.pipelines.list(status=status,  # type: ignore[assignment]
+                                                                             ref=branch_name,
+                                                                             source=triggering_source)
+        for pipeline in pipelines_for_status:
+            pipelines[pipeline.id] = pipeline
+    return list(pipelines.values())
+
+
+def cancel_pipelines_for_branch_name(gitlab_client: Gitlab,
+                                     project: Project,
+                                     branch_name: str,
+                                     triggering_source: str,
+                                     cancel_children: bool = True,
+                                     pipeline_id: int | None = None) -> bool:
+
+    pipelines = get_all_pipelines_for_all_statuses(project, branch_name, triggering_source)
+    if not pipelines:
+        logging.info(f"No pipelines found for branch:{branch_name}, triggering source:{triggering_source}")
+        return True
+
+    logging.info(f"Canceling {len(pipelines)} pipelines for branch:{branch_name}, triggering source:{triggering_source}")
+    success = True
+    for pipeline in pipelines:
+        # Only cancel pipelines that were created before the current pipeline, If there is a newer
+        # pipeline it will cancel our run, If the build number is None cancel all pipelines.
```

Comment:
Can't there be a race condition if the build number is higher than ours?

Reply:
if both pipelines are running at the same time, the one with the higher number will win. that's why we've this condition.